### ISSUE TITLE
API proposal: Unsafe.AlignOf<T>()

### ISSUE THREAD

**Issue by GrabYourPitchforks** [dotnet/runtime#2167]
Forked from a conversation at https://github.com/dotnet/corefx/issues/26465.

API proposal:

```cs
namespace System.Runtime.CompilerServices
{
    public static class Unsafe
    {
        public static int AlignOf<T>();
    }
}
```

The behavior of this API would be that it returns the natural alignment (in bytes) of the data type `T`. This is most useful for components which need to *reinterpret_cast* between two different types `T` and `U` and wish to check ahead of time whether alignment problems might result. See further discussion of this scenario at https://github.com/dotnet/corefx/issues/26465.

**Comment by tannergooding** [comment]
Would this be the heap alignment or stack alignment of the type (as these sometimes differ)?

What would the consideration be when the native alignment of a type and the heap alignment differ (e.g. `Vector128<T>` is 16-byte aligned on the stack and in native, but is 8-byte aligned for the heap)?

**Comment by GrabYourPitchforks** [comment]
I would assume heap-alignment, since that's the worst-case scenario where a VM or compiler might naturally put a field of type `T` insider of a larger object. That's assuming I'm understanding the scenario correctly.

**Comment by jkotas** [comment]
`System.Runtime.CompilerServices.Unsafe` is runtime agnostic today. This API cannot be implemented in runtime agnostic way. Adding it to `System.Runtime.CompilerServices.Unsafe` would mean killing it as runtime agnostic package. Should be rather be somewhere else?

> understanding the scenario correctly.

Yup, there is a lot of different values that this method can return depending on what it is meant to be  used for...

**Comment by GrabYourPitchforks** [comment]
In ECMA-335, Sec. I.12.6.2, the below clause appears.

> Otherwise, if the data is not properly aligned, and no `unaligned.` prefix has been specified, executing the instruction can generate unaligned memory faults or incorrect data.

If we wanted to use this sentence as a guide, then the return value of `AlignOf<T>()` might be defined as "the address alignment value at which accessing a pointer of type `T*` will not generate unaligned memory faults or incorrect data on the current architecture, even if the instruction has not been decorated with an `unaligned.` prefix."

Over at https://github.com/dotnet/corefx/issues/26368#issuecomment-358437182 (which eventually led to this proposal) there's discussion about some architectures possibly generating faults when trying to access unaligned data. But TBH I'm not familiar enough with the silicon to speak to that.

**Comment by jkotas** [comment]
There is a number of cases where misaligned `T*` will not generate fault, but accessing it comes with performance penalty. Would any penalty be acceptable? 1.5x? 5x? 15x?

**Comment by ericstj** [comment]
Given the linked issue was moved to future, and the timing for 5.0, will also move this to future as well.  @GrabYourPitchforks feel free to move as you see fit.

Also given @jkotas comment about this being runtime-specific, my inclination would be to put this in a System.Memory-related-type instead of Unsafe.

**Comment by GrabYourPitchforks** [comment]
Agree on both counts. Though perhaps if it's runtime-dependent and meant for interop, then `System.Runtime.InteropServices` would be preferred over `System.Memory`?

**Comment by ericstj** [comment]
Yeah, I considered that too (EG: Marshal) however I got the impression it was useful for more scenarios than interop.  EG: Intrinsics and optimizing perf might care about alignment.

**Comment by xtqqczze** [comment]
In absence of this API, there is a workaround:

```csharp
[StructLayout(LayoutKind.Sequential)]
private struct AlignOfHelper<T> where T : unmanaged
{
    public byte dummy;
    public T data;
}
public unsafe static int AlignOf<T>() where T : unmanaged
{
    return sizeof(AlignOfHelper<T>) - sizeof(T);
}
```

Adapted from: https://github.com/Unity-Technologies/UnityCsReference/blob/e7d9de5f09767c3320b6dab51bc2c2dc90447786/Runtime/Export/Unsafe/UnsafeUtility.cs#L121-L132

**Comment by tannergooding** [comment]
@jkotas is the runtime agnostic consideration still applicable today? This seems more doable in .NET 12

If not or we think its not reasonable enough to provide, then we should just close it. The main benefit is that other ecosystems provide it and if we make the decision to report back the "natural packing" (i.e. native alignment, rather than GC alignment) then users can pair it with `NativeMemory` (which `NativeMemory.AlignOf` may be a better location accordingly).

**Comment by jkotas** [comment]
> is the runtime agnostic consideration still applicable today? 

Right, it does not apply anymore since we have moved S.R.CS.Unsafe inbox.

> users can pair it with NativeMemory (which NativeMemory.AlignOf may be a better location accordingly).

Yes, `Unsafe` may not be the best place for this API depending on what the API actually does.
